x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/oidc/OpenIdConnectAuthenticator.java

```diff
@@ -670,8 +670,18 @@ static JSONObject mergeObjects(JSONObject idToken, JSONObject userInfo) {
             } else if (value1 instanceof JSONObject) {
                 idToken.put(entry.getKey(), mergeObjects((JSONObject) value1, value2));
             } else if (value1.getClass().equals(value2.getClass()) == false) {
-                throw new IllegalStateException("Error merging ID token and userinfo claim value for claim [" + entry.getKey() + "]. " +
-                    "Cannot merge [" + value1.getClass().getName() + "] with [" + value2.getClass().getName() + "]");
+                // A special handling for certain OPs that mix the usage of true and "true"
+                // Retain value from idToken as all other primitive types
+                if (value1 instanceof Boolean && value2 instanceof String
+                    && ("true".equalsIgnoreCase((String)value2) || "false".equalsIgnoreCase((String)value2))) {
```

Comment:
I wonder if we should care whether they're the same logical value? We don't for other cases, but I think we should be.
Perhaps:
```suggestion
                    && String.valueOf(value1).equals(value2) ) {
```

Reply:
Updated as suggested. It's a bit inconsistent with other primitive types. But I think it fits for "minimal leniency" purpose which is I am after currently. Also, as discussed, it's worthwhile to revisit the whole merging behaviour in future.